### PR TITLE
Add Mapper memoization

### PR DESCRIPTION
**Pull request by henryre**
Adds memoization to `BaseMapper` classes (preprocessor, TF, etc.).

* BaseMapper objects just maintain a simple dictionary cache which is populated/read from if memoization is turned on
* Only complexity is in `get_hashable`, which recursively generates a hashable value type for common unhashable data types we encounter (SimpleNamespace, pd.Series, np.ndarray, list, dict). It's a bit tricky, so it's extensively tested in both `test/labeling/apply/test_lf_applier.py` and `test/map/test_core.py`
* Considered using `functools.lru_cache`, but it doesn't handle unhashable types
* `lambda_mapper` is now a class so it has parens (@bhancock8 @vincentschen)

There are times we could manually reset the preprocessor caches (e.g. LFApplier.__call__), but not sure we always want to. Opinions welcome. In a follow-up diff, will introduce nested mappers. These two diffs together mean we avoid re-execution of preprocessors when they're used by multiple LFs.

**Test plan**
* Lots of added unit tests for `get_hashable` and BaseMapper caching
* Speed-up test results below. Ran on 2000k English movie review sentences with 4 LFs that use SpaCy. Speed-up is close to 4x (corresponding to reduction in SpaCy calls).

Screen Shot 2019-06-28 at 7 23 19 PM

### CODE REVIEW

File: snorkel/map/core.py [hunk 1]
```diff
@@ -28,19 +33,90 @@ def get_parameters(
     return params[0]
 
 
+def is_hashable(obj: Any) -> bool:
+    """Test if object is hashable via duck typing.
+
+    NB: not using `collections.Hashable` as some objects
```
Comment (bhancock8) [inline snorkel/map/core.py]: NB?

Comment (bhancock8) [reply snorkel/map/core.py]: By usage, I assume this just means "Note:", but the latter seems clearer?

Comment (bhancock8) [reply snorkel/map/core.py]: (Whatever we decide, update all instances of NB)

File: test/map/test_core.py [hunk 2]
```diff
@@ -36,29 +41,41 @@ def run(self, text: str, **kwargs: Any) -> Optional[FieldMap]:  # type: ignore
...
+    def _get_x(self, num=8, text="Henry has fun") -> SimpleNamespace:
+        return SimpleNamespace(num=num, text=text)
 
     def _get_x_dict(self) -> SimpleNamespace:
         return SimpleNamespace(num=8, d=dict(my_key=1))
 
     def test_numeric_mapper(self) -> None:
         square.set_mode(MapperMode.NAMESPACE)
         x_mapped = square(self._get_x())
+        # NB: not using `self.assertIsNotNone` due to mypy
```
Comment (bhancock8) [inline test/map/test_core.py]: Could we do `self.assertTrue(x_mapped is not None)` instead?

File: snorkel/map/core.py (hunk 1 above)
Comment (henryre) [reply snorkel/map/core.py]: This is going to be so pretentious. https://en.wikipedia.org/wiki/Nota_bene. It means "Important note". Pretty common in code comments.

File: test/map/test_core.py (hunk 2 above)
Comment (henryre) [reply test/map/test_core.py]: Hmmm, I'll try it

File: snorkel/map/core.py (hunk 1 above)
Comment (ajratner) [reply snorkel/map/core.py]: Hahahaha I literally read Braden's first comment, and was like- I can't wait for Henry's response it's going to be great ("Well, on the Yale Daily News, we used it for...")

Comment (ajratner) [reply snorkel/map/core.py]: However I also have seen it a lot fwiw, although I also default to 'Note:'

File: snorkel/map/core.py [hunk 3]
```diff
@@ -28,19 +33,90 @@ def get_parameters(
...
+    a more specific exception.
+    """
+    try:
+        hash(obj)
+        return True
+    except Exception:
+        return False
+
+
+def get_hashable(obj: Any) -> Hashable:
```
Comment (ajratner) [inline snorkel/map/core.py]: FWIW: If we did the approach where we converted every data type to a common IR, we could just make sure that said IR was hashable with `hash` and skip this complexity, which makes me nervous...

File: snorkel/map/core.py [hunk 4]
```diff
@@ -28,19 +33,90 @@ def get_parameters(
...
+
+    def memoize_outputs(self, mode: bool) -> None:
+        """Set memoization mode.
+
+        Parameters
+        ----------
+        mode
+            Memoize mapper outputs?
+        """
+        self._memoize_outputs = mode
```
Comment (ajratner) [inline snorkel/map/core.py]: Is this a little overkill?  Can't we just set the `self._memoize_outputs` attrib in init...?  I *really* want us to avoid code complexity that isn't strictly needed... it is going to *kill* us at this stage.

Regardless, if we move to only using getters / setters a la Java, I think we should be consistent throughout, and I don't think we're anywhere near there...

File: snorkel/map/core.py [hunk 5]
```diff
@@ -52,7 +128,10 @@ def set_mode(self, mode: MapperMode) -> None:
         """
         self.mode = mode
```
Comment (ajratner) [inline snorkel/map/core.py]: Ditto here re: comment above

File: snorkel/map/core.py (hunk 3 above)
Comment (ajratner) [reply snorkel/map/core.py]: (this is not a minor comment, but not an actionable one in this PR...)

Comment (henryre) [reply snorkel/map/core.py]: Not necessarily...the values in certain fields of the IR might not be hashable (e.g. a dictionary). I agree this would help with the top-layer, but we'd still need this functionality. Will keep it on the roadmap.

File: snorkel/map/core.py (hunk 4 above)
Comment (henryre) [reply snorkel/map/core.py]: ☠️

Of course we're not going towards setters and getters, but I do see people preferring this type of mode setting (e.g. `torch.nn.Module.train(...)`) for things that actual users interact with (not just other classes). Open to just making `memoize` a public attrib for now.

File: snorkel/map/core.py (hunk 5 above)
Comment (henryre) [reply snorkel/map/core.py]: Ooooh you shouldn't have highlighted this one...now I want to leave these methods for proper type checking.

Comment (henryre) [reply snorkel/map/core.py]: Scratch that, mypy will yell if we type the attribute.

File: snorkel/map/core.py (hunk 4 above)
Comment (henryre) [reply snorkel/map/core.py]: Although `train(...)` sets train mode recursively, so that's a bad example. Ok let's make it a public attrib

File: test/map/test_core.py (hunk 2 above)
Comment (henryre) [reply test/map/test_core.py]: Answer: nope

**Comment by henryre** [comment]
Travis failure due to non-deterministic slicing test, merging
